contributors/design-proposals/admission_control_extension.md

```diff
@@ -0,0 +1,594 @@
+for extending the core with additional policy. This prevents the formation of an open ecosystem for tools
+*around* Kubernetes, forcing all changes to policy to go through the Kubernetes codebase review gate (when
+such review is unnecessary or disruptive to Kubernetes itself).
+
+
+## Design
+
+It should be possible to perform holistic policy enforcement in Kubernetes without the recompilation of the
+core project as plugins that can be added and removed to a stock Kubernetes release. That extension
+of admission control should leverage similar mechanisms to our existing controller frameworks where possible
+and otherwise be performant and reliable.
+
+
+### Requirements
+
+1.  Easy Initialization
+
+    Privileged components should be able to easily participate in the **initialization** of a new object.
+
+2.  Synchronous Validation
+
+    Synchronous rejection of initialized objects or mutations must be possible outside of the kube-apiserver binary
+
+3.  Backwards Compatible
+
+    Existing API clients must see no change in behavior to external admission other than increased latency
+
+4.  Easy Installation
+
+    Administrators should be able to easily write a new admission plugin and deploy it in the cluster
+
+5.  Performant
+
+    External admission must not significantly regress performance in large and dense clusters
+
+6.  Reliable
+
+    External admission should be capable of being "production-grade" for deployment in an extremely large and dense cluster
+
+7.  Internally Consistent
+
+    Developing an admission controller should reuse as much infrastructure and tools as possible from building custom controllers so as to reduce the cost of extension.
+
+
+### Specification
+
+Based on observation of the actual admission control implementations the majority of mutation
+occurs as part of creation, and a large chunk of the remaining controllers are for side-effect free
+validation of creation and updates. Therefore we propose the following changes to Kubernetes:
+
+1.  Allow some controllers to act as "initializers" - watching the API and mutating the object before it is visible to normal clients.
+
+    This would reuse the majority of the infrastructure in place for controllers. Because creation is
+    one-way, the object can be "revealed" to regular clients once a set list of initializers is consumed. These
+    controllers could run on the cluster as pods. Because initialization is a non-backwards compatible API change,
+    some care must be taken to shield old clients from observing the scenario.
+
+2.  Add a generic **external admission webhook** controller that is non-mutating (thus parallelizable)
+
+    This generic webhook API would resemble `admission.Interface` and be given the input object (for create) and the
+    previous object (for update). After initialization or on any update, these hooks would be invoked in parallel
```

Comment:
Andy had a great write up of the logical steps of admission:

1. defaulting
2. mutating complex defaulting
3. validation
4. quota or other admission that must mutate another resource and potentially roll back

If I had to straw man a "total order" it might be

1. normal admission chain defaulting (PodPreset)
2. initializer admission (just a form of defaulting with a touch of security)
3. external mutating (now or later)
4. other security / validating (PodSecurityPolicy)
5. external admission, non-mutating 
6. quota: object count quota for uninitialized resources, full quota for initialized resources

I think both external admission and initializers are implemented as new admission plugins - the external admission plugin could appear multiple times in the chain (mutating and non-mutating), some "points" for external admission in the chain might evolve over time (post quota?).

I'll add a diagram to the doc that includes a hypothetical flowchart of all known admission plugins.